File: leet_Design_0208_trie.py
class TrieNode:
    def __init__(self):
        self.branch = defaultdict(TrieNode)
        self.ended = False

class Trie:

    def __init__(self):
        self.node = TrieNode()

    def insert(self, word: str) -> None:
        cur = self.node
        for c in word:
            if c not in cur.branch:
                cur.branch[c] = TrieNode()
            cur = cur.branch[c]
        cur.ended = True

    def search(self, word: str) -> bool:
        cur = self.node
        for c in word:
            if c not in cur.branch:
                return False
            cur = cur.branch[c]
        return cur.ended

    def startsWith(self, prefix: str) -> bool:
        cur = self.node
        for c in prefix:
            if c not in cur.branch:
                return False
            cur = cur.branch[c]
        return True


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: leet_Design_0208_trie.go
type TrieNode struct {
    ended bool
    branch [26] * TrieNode
}

type Trie struct {
    node * TrieNode
}


func Constructor() Trie {
    return Trie {
        node: & TrieNode {},
    }
}


func (this *Trie) Insert(word string)  {
    cur := this.node
    for _, c := range word {
        if cur.branch[c - 'a'] == nil {
            cur.branch[c - 'a'] = & TrieNode {}
        }
        cur = cur.branch[c - 'a']
    }
    cur.ended = true
}


func (this *Trie) Search(word string) bool {
    cur := this.node
    for _, c := range word {
        if cur.branch[c - 'a'] == nil {
            return false
        }
        cur = cur.branch[c - 'a']
    }
    return cur.ended 
}


func (this *Trie) StartsWith(prefix string) bool {
    cur := this.node
    for _, c := range prefix {
        if cur.branch[c - 'a'] == nil {
            return false
        }
        cur = cur.branch[c - 'a']
    }
    return true
}


/**
 * Your Trie object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Insert(word);
 * param_2 := obj.Search(word);
 * param_3 := obj.StartsWith(prefix);
 */
